core/transcoder.go

```diff
@@ -55,34 +56,92 @@ func NewLocalTranscoder(workDir string) Transcoder {
 	return &LocalTranscoder{workDir: workDir}
 }
 
+type nvSegResult struct {
+	*TranscodeData
+	error
+}
+
+type nvSegData struct {
+	session  *ffmpeg.Transcoder
+	fname    string
+	profiles []ffmpeg.VideoProfile
+	res      chan *nvSegResult
+}
+
 type NvidiaTranscoder struct {
-	workDir string
-	device  string
+	device  *segStack
 	session *ffmpeg.Transcoder
 }
 
-func (nv *NvidiaTranscoder) Transcode(job string, fname string, profiles []ffmpeg.VideoProfile) (*TranscodeData, error) {
-	// Set up in / out config
-	in := &ffmpeg.TranscodeOptionsIn{
-		Fname:  fname,
-		Accel:  ffmpeg.Nvidia,
-		Device: nv.device,
+type segStack struct {
+	segs []*nvSegData
+	mu   *sync.Mutex
+	cv   *sync.Cond
+	gpu  string
+}
+
+func newSegStack() *segStack {
+	mu := &sync.Mutex{}
+	return &segStack{
+		segs: []*nvSegData{},
+		mu:   mu,
+		cv:   sync.NewCond(mu),
 	}
-	opts := profilesToTranscodeOptions(nv.workDir, ffmpeg.Nvidia, profiles)
+}
 
-	// Do the Transcoding
-	res, err := nv.session.Transcode(in, opts)
-	if err != nil {
-		return nil, err
+func (ss *segStack) push(seg *nvSegData) {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+	ss.segs = append(ss.segs, seg)
+	ss.cv.Broadcast()
+	if monitor.Enabled {
+		monitor.GPUBacklog(ss.gpu, len(ss.segs))
 	}
+}
 
-	return resToTranscodeData(res, opts)
+func (ss *segStack) pop() *nvSegData {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+	for len(ss.segs) <= 0 {
+		ss.cv.Wait()
+	}
+	seg := ss.segs[len(ss.segs)-1]
+	ss.segs = ss.segs[:len(ss.segs)-1]
+	if monitor.Enabled {
+		monitor.GPUBacklog(ss.gpu, len(ss.segs))
+	}
+	return seg
+}
+
+func (nv *NvidiaTranscoder) Transcode(job string, fname string, profiles []ffmpeg.VideoProfile) (*TranscodeData, error) {
+
+	segData := &nvSegData{
+		session:  nv.session,
+		fname:    fname,
+		profiles: profiles,
+		res:      make(chan *nvSegResult, 1),
+	}
+	nv.device.push(segData)
+	res := <-segData.res
+	return res.TranscodeData, res.error
+}
+
+var nvidiaDevices map[string]*segStack
+
+func StartNvidiaTranscoders(devices string, workDir string) {
+	nvidiaDevices = make(map[string]*segStack)
+	d := strings.Split(devices, ",")
+	for _, n := range d {
+		st := newSegStack()
+		st.gpu = n
```

Comment:
Could `gpu` be set in `newSegStack()`?

Reply:
Fixed in 26c1f19b46bb9546bdbce22cbe3f890d957e0310